go/client/chat_resolver.go

```diff
@@ -7,105 +7,202 @@ import (
 	"context"
 	"errors"
 	"fmt"
-	"strconv"
 
 	"github.com/keybase/client/go/libkb"
 	"github.com/keybase/client/go/protocol/chat1"
 	"github.com/keybase/client/go/protocol/keybase1"
 )
 
-type chatConversationResolvingRequestContext struct {
-	canonicalizedTlfName string
+type NewConvWithAliasError struct{}
+
+func (e NewConvWithAliasError) Error() string {
+	return "cannot create conversation using only conversation alias"
 }
 
-type chatConversationResolvingRequest struct {
-	TlfName    string
-	TopicName  string
-	TopicType  chat1.TopicType
-	Visibility chat1.TLFVisibility
+type NewConvWithoutTLFNameError struct{}
 
-	ctx *chatConversationResolvingRequestContext
+func (e NewConvWithoutTLFNameError) Error() string {
+	return "cannot create conversation without a TLF name"
 }
 
-type chatConversationResolvingBehavior struct {
-	// Specify whether the resolver should use CLI prompts to ask for more
-	// information from user. If false, Resolve errors when multiple conversation
-	// is found, or when trying to create a new conversation without enough
-	// information.
-	Interactive bool
+type NoConversationFoundError struct{}
 
-	// Specify whether the resolve should create a conversation if none is found.
-	CreateIfNotExists bool
+func (e NoConversationFoundError) Error() string {
+	return "no conversation found"
 }
 
-type chatConversationResolver struct {
-	G          *libkb.GlobalContext
-	ChatClient chat1.LocalInterface
-	TlfClient  keybase1.TlfInterface
+type MultipleConversationFoundError struct{}
+
+func (e MultipleConversationFoundError) Error() string {
+	return "multiple conversation found"
+}
+
+type ChatTLFNameCookerRequest struct {
+	Raw      string
+	IsPublic bool
 }
 
-// completeAndCanonicalizeTLFName completes tlfName and canonicalizes it if
-// necessary. The new TLF name is stored to req.ctx.canonicalizedTlfName.
-// len(tlfName) must > 0
-// TODO: support public TLFs
-func (r *chatConversationResolver) completeAndCanonicalizeTLFName(
-	ctx context.Context, tlfName string, req chatConversationResolvingRequest) error {
-	cname, err := r.TlfClient.CompleteAndCanonicalizePrivateTlfName(ctx, keybase1.TLFQuery{
-		TlfName:          tlfName,
+type ChatTLFNameCooker interface {
+	Cook(ctx context.Context, req ChatTLFNameCookerRequest) (cooked string, err error)
+}
+
+type chatCLITLFNameCooker struct {
+	tlfClient  keybase1.TlfInterface
+	terminalUI libkb.TerminalUI
+
+	hippocampus map[ChatTLFNameCookerRequest]string
```

Comment:
Sorry to be against fun, but I think something more standard would be good in the long run.